aot-maven-plugin/src/main/java/com/dylibso/chicory/experimental/maven/aot/AotGenMojo.java

```diff
@@ -123,12 +129,126 @@ public void execute() throws MojoExecutionException {
+
+        cu.addImport(Instance.class);
+        cu.addImport(Machine.class);
+
+        var clazz = cu.addClass(machineFactoryName, Keyword.PUBLIC, Keyword.FINAL);
+
+        var constr = clazz.addConstructor(Keyword.PRIVATE);
+        constr.createBody();
+
+        var method = clazz.addMethod("create", Keyword.PUBLIC, Keyword.STATIC);
+        method.addParameter(parseType("Instance"), "instance");
+        method.setType(Machine.class);
+        var methodBody = method.createBody();
+
+        var constructorInvocation =
+                new ObjectCreationExpr(
+                        null,
+                        parseClassOrInterfaceType(machineName),
+                        NodeList.nodeList(new NameExpr("instance")));
+        methodBody.addStatement(new ReturnStmt(constructorInvocation));
+
+        dest.add(packageName, machineFactoryName + ".java", cu);
+    }
+
+    private static void generateModule(
+            SourceRoot dest, String packageName, String moduleName, String wasmName) {
+        var cu = new CompilationUnit(packageName);
+
+        cu.addImport(IOException.class);
+        cu.addImport(UncheckedIOException.class);
+        cu.addImport(Parser.class);
+        cu.addImport(WasmModule.class);
+
+        var clazz = cu.addClass(moduleName, Keyword.PUBLIC, Keyword.FINAL);
+
+        var constr = clazz.addConstructor(Keyword.PRIVATE);
+        constr.createBody();
+
+        var method = clazz.addMethod("load", Keyword.PUBLIC, Keyword.STATIC);
+        method.setType(WasmModule.class);
+        var methodBody = method.createBody();
+
+        var getResource =
+                new MethodCallExpr(
+                        new ClassExpr(parseType(moduleName)),
+                        "getResourceAsStream",
+                        new NodeList<>(new StringLiteralExpr(wasmName)));
+        var resourceVar =
+                new VariableDeclarationExpr(
+                        new VariableDeclarator(new VarType(), "in", getResource));
+
+        var returnStmt =
+                new ReturnStmt(
+                        new MethodCallExpr()
+                                .setScope(new NameExpr("Parser"))
+                                .setName("parse")
+                                .addArgument(new NameExpr("in")));
+
+        var newException =
+                new ObjectCreationExpr()
+                        .setType(parseClassOrInterfaceType("UncheckedIOException"))
+                        .addArgument(new StringLiteralExpr("Failed to load AOT WASM module"))
+                        .addArgument(new NameExpr("e"));
+        var catchIoException =
+                new CatchClause()
+                        .setParameter(
+                                new com.github.javaparser.ast.body.Parameter(
+                                        parseClassOrInterfaceType("IOException"), "e"))
+                        .setBody(new BlockStmt(new NodeList<>(new ThrowStmt(newException))));
```

Comment:
Generated code should not catch Exception unless in specific cases.
I would just add the Exception to the signature.

Reply:
This is try-with-resources, which is needed to close the `Class.getResourceAsStream()` stream. This follows the model of `Parser.parse()` in not throwing a checked exception.

From a practical standpoint, making `load()` throw `IOException` would be annoying, as it would prevent this type of usage, requiring a static initializer block:
```java
private static final WasmModule MODULE = Wast2JsonModule.load();
```

Changing `Parser` makes more sense, as that can be used with arbitrary resources. This code is used with a resource in the same JAR, so any failure would be quite unexpected, and not something that user code is expected to handle (which is the typical reasoning for checked exceptions).

Reply:
Thanks for the explanation, convinced :+1: